### ISSUE TITLE
Refactor code + use new exposes info provided by zigbee2mqtt

### ISSUE THREAD

**Issue by itavero** [itavero/homebridge-z2m#18]
I noticed `platformAccessory.ts` is getting very long (it's over 900 lines).
Some of the interfaces should move to separate sources, as well as all the different `ServiceWrapper` implementations.

**Comment by itavero** [comment]
I've also noticed that some keys can apply to multiple services.
For instance, several CO/gas/smoke detectors also have a tamper detection and some of them to for instance both gas and CO.
This means that the `tamper` key, if present, can be used for an optional [Status Tampered](https://developers.homebridge.io/#/characteristic/StatusTampered) characteristic in both the _Carbon Monoxide Sensor_ and the _Leak Sensor_ service.

With the current design/implementation this is however not possible.

I think I need to keep track of which keys have been used, but still give other ServiceWrappers also the possibility to use them.
And ofcourse, I probably also need an additional abstraction to reuse the code for these "reusable" characteristics that can be used by multiple Services.

**Comment by itavero** [comment]
For the refactoring, I'm also waiting on the new `exposes` attribute to be released for zigbee2mqtt (see Koenkk/zigbee2mqtt#4466). This also means that I'm "postponing" the major release.

Once implemented, users will be required to use a certain minimum version of zigbee2mqtt and have the new api enabled.
This might be a bit of additional work for users to updated and get things working again, but I think it outways the benefits.
This plugin will have a far better out of the box experience, as it will be better at detecting the right mapping and it will do so on startup, instead of having to wait for the right data to pass by.

**Comment by itavero** [comment]
The mentioned attribute is in the latest release, so I'll start refactoring code and mechanisms soon.

**Comment by itavero** [comment]
Currently busy implementing this in the `z2m-exposes` branch.
Already implemented support for lights. I'll continue with the rest in the coming weeks.
My goal is to have a new version out before the end of the year, that supports at least the devices it supports right now, but via the new `exposes` information.

I'll try to post updates in this issue while I'm implementing.

**Comment by itavero** [comment]
Short update: currently I have the lights and some sensors working.
I already noticed a few things that I want to clean up though (one of the "problems" of writing code is that you gain new insights while doing it, which often lead to other ideas and therefor to refactoring code you've just written). 😅

I also want to add unit tests to verify the implementations (so that the regression tests become automated and reproducable, instead of me just running the plugin and trying some stuff at random). I might add the unit tests later and focus on getting the currently supported devices in there first, so I can put a "pre-release" out there for people to try out.

**Comment by itavero** [comment]
Did some more work on the new version this evening.
In general the internal APIs and architecture are still fluid, although I think they are getting more stable.

Current state of what has been implemented and what not (based on the services mentioned in the README of the current version):
- [x] Switch
- [x] Lightbulb
- [ ] Lock Mechanism
- [x] Temperature Sensor
- [x] Humidity Sensor
- [x] Contact Sensor
- [x] Light Sensor
- [x] Smoke Sensor
- [x] Leak Sensor (water)
- [x] Leak Sensor (gas)
- [ ] Window Covering
- [ ] Battery Service
- [x] Carbon Monoxide Sensor
- [ ] Air Pressure Sensor (based on Elgato Eve Weather characteristics)
- [ ] Energy metering stuff (based on Elgato Eve Energy characteristics)

Even though I might have checked of certain services in the list above, decent testing/verification is still required for most of them.
Also, as mentioned, I still want to write automated tests to prove the conversions and such are setup correctly and taken over from the `exposes` information provided by zigbee2mqtt correctly.

Next up will probably be the _Window Covering_ as that is one of the most used accessories in my home (after the Lightbulbs of course).

With regards to the _Lock Mechanism_, I don't have an actual device to verify it with, so I can only do "theoretical" verification based on the code of [Koenkk/zigbee-herdsman-converters](https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/lib/exposes.js).

**Comment by itavero** [comment]
Just implemented Window Covering. (updated the previous comment)

**Comment by itavero** [comment]
Just implemented the Battery Service. (updated the comment above)

Next up will probably be the Air Pressure Sensor or starting on the unit tests.

**Comment by jwilling** [comment]
I've been following along with your progress on this and it's looking amazing! Can't wait to try it out once it's released.

**Comment by itavero** [comment]
@jwilling Thanks for your message and your support. It really means a lot to me. 👍

It feels like I'm still on track for making a pre-release before the end of the year.

The main "risk" will be in the automated tests, as I've never wrote tests in Typescript before and there aren't a lot of resources specifically on testing homebridge plugins.
But I'll figure it out, one way or the other 😉

**Comment by itavero** [comment]
Implemented Air Pressure Sensor. (updated list from earlier comment)

Tried to start on the Lock Mechanism, but up on diving into the zigbee-herdsman-converters I noticed there isn't a single lock exposing the `lock_state`, so I'm not sure how reliable the exposes data for the locks is at the moment.

**Comment by itavero** [comment]
`lock_state` exposes have been added just now in zigbee-herdsman-converters. Next time I have some time left, I'll probably implement the Lock Mechanism, followed by the Energy measurement stuff.

After that I'm planning to dive into the automated tests.

**Comment by itavero** [comment]
I tried working on the Lock Mechanism converter/handler yesterday, but I noticed that the `lock_state` isn't part of the `lock` type's features, which I would expect.
Proposed this change in Koenkk/zigbee-herdsman-converters#1930.

Just now I've added some documentation on the implemented services/characteristics (in the `docs` folder).

**Comment by itavero** [comment]
PR fo zigbee-hersmand-converters got merged, so I've implemented the Lock Mechanism handler. (I've updated the list above).

This weekend I hope to implement the energy metering stuff. If I succeed, it should be doable to publish a "pre release" version before Christmas.

A pre-release would allow users to give the new version a try and give feedback, while I'm still doing more verification using automated tests.

**Comment by itavero** [comment]
I've decided to leave the energy metering stuff out. I personally did not have a use case for this information, so I figured I'd better spend my time and effort on other stuff I do actually use / care about. If you miss this, please open up a feature request and also let me know what this information is useful for within a HomeKit context.

**Comment by itavero** [comment]
A pre-release before Christmas won't happen.

So far I've only written a few tests for the Lightbulb and Switch handlers and I've already found a couple of silly mistakes. I feel I should do some more automated tests before publishing it, just to gain some more confidence. 😅

Besides that, this week I had to (read: wanted to) spent a bit more time helping my wife with her hobby (🐴 🦙) than I anticipated.

I still think a pre-release before the year is over is doable. Next week I should have plenty of time to wrap things up.

**Comment by itavero** [comment]
Added more tests today for the "basic" sensor devices.
Also found and fixed some bugs along the way.

The following services have no automated tests (yet):
* Battery Service
* Lock (but I don't have an actual device myself to base the tests on)
* Window Covering (tricky one given there is some more logic in it to get the status updates)

I'll have a look at the coverage report tomorrow to see if I didn't miss any important stuff in the other tests.
The current test setup for Lights / Switches can still be improved as the tests currently slightly depend on each other, which of course is not a good practice when writing tests.

I will probably also go over all the changes one more time, just to see if I still see some things that I need to polish before making a release. 😉

**Comment by itavero** [comment]
Added some more tests today (mainly for the Window Covering) and made some additional changes.

I plan to merge #31 tomorrow and make a release available under `homebridge-z2m@next` after that, which I will be installing on my "production" environment at home as well.

**Comment by itavero** [comment]
Version 1.0.2 is now available via `npm install homebridge-z2m@next`.
I'm currently running it at home.

Note that all accessories will be recreated at this point. I might have a look if I can improve on this by just removing all the services instead, before recreating them.

**Comment by tkrtz** [comment]
The newest version doesn’t seem to recognize my devices. Currently I‘m using the newest version (1.16.2) of z2m. Do I still need to do anything else to make it work?

**Comment by itavero** [comment]
@tkrtz You should not need to do anything else.
Can you share your logs maybe?

Which version of MQTT are you running?
Do you see a message on the `zigbee2mqtt/bridge/devices` topic?

**Comment by tkrtz** [comment]
I‘m using the latest version of mosquitto and there sadly isn‘t a topic called `zigbee2mqtt/bridge/devices`.
Which log do you want to see?

**Comment by itavero** [comment]
Have you also configured zigbee2mqtt and homebridge-z2m to use MQTT version 5?
I think it might be because zigbee2mqtt publishes the message to `zigbee2mqtt/bridge/devices` as retained, but now I'm doubting if this would work if the MQTT version is 4 or earlier.

I'll browse through the code of zigbee2mqtt.

I'm mainly interested in the logs of homebridge when it starts, it should list the accessories it is creating or restoring.

**Comment by tkrtz** [comment]
Okay for MQTT version 4:
`^[[37m[31/12/2020, 16:12:45]^[[0m ^[[36m[HB Supervisor]^[[0m Restarting Homebridge...
^[[37m[31/12/2020, 16:12:45]^[[0m ^[[36m[HB Supervisor]^[[0m Starting Homebridge with extra flags: -I
^[[37m[31/12/2020, 16:12:45]^[[0m ^[[36m[HB Supervisor]^[[0m Started Homebridge v1.1.7 with PID: 5104
^[[37m[31/12/2020, 16:13:03]^[[0m ^[[36m[HB Supervisor]^[[0m OS: Linux 5.4.79+ arm
^[[37m[31/12/2020, 16:13:03]^[[0m ^[[36m[HB Supervisor]^[[0m Node.js v15.5.0 /usr/local/bin/node
^[[37m[31/12/2020, 16:13:03]^[[0m ^[[36m[HB Supervisor]^[[0m Homebridge Path: /usr/local/lib/node_modules/homebridge/bin/homebridge
^[[37m[31/12/2020, 16:13:03]^[[0m ^[[36m[HB Supervisor]^[[0m UI Path: /usr/local/lib/node_modules/homebridge-config-ui-x/dist/bin/s$^[[37m[31/12/2020, 16:13:03]^[[0m ^[[36m[HB Supervisor]^[[0m Delaying Homebridge startup by 20 seconds on low powered server
^[[37m[31/12/2020, 16:13:24]^[[0m ^[[36m[HB Supervisor]^[[0m Starting Homebridge with extra flags: -I
^[[37m[31/12/2020, 16:13:24]^[[0m ^[[36m[HB Supervisor]^[[0m Started Homebridge v1.1.7 with PID: 5134
^[[0;37m[31/12/2020, 16:13:47] ^[[0m^[[0;36m[Homebridge UI]^[[0m ^[[0;33mHomebridge Config UI X v4.36.0 is listening on :: port 858$^[[37m[31/12/2020, 16:13:37] ^[[39mLoaded config.json with 0 accessories and 3 platforms.
^[[37m[31/12/2020, 16:13:38] ^[[39m---
^[[37m[31/12/2020, 16:13:53] ^[[39mLoaded plugin: homebridge-config-ui-x@4.36.0
^[[37m[31/12/2020, 16:13:53] ^[[39mRegistering platform 'homebridge-config-ui-x.config'
^[[37m[31/12/2020, 16:13:53] ^[[39m---
^[[37m[31/12/2020, 16:13:53] ^[[39mLoaded plugin: homebridge-intercom-automation-hat@1.0.3
^[[37m[31/12/2020, 16:13:53] ^[[39mRegistering platform 'homebridge-intercom-automation-hat.IntercomAutomationHAT'
^[[37m[31/12/2020, 16:13:53] ^[[39mRegistering accessory 'homebridge-intercom-automation-hat.IntercomAutomationHATLockAccessory'
^[[37m[31/12/2020, 16:13:53] ^[[39mRegistering accessory 'homebridge-intercom-automation-hat.IntercomAutomationHATBellAccessory'
^[[37m[31/12/2020, 16:13:53] ^[[39m---
^[[37m[31/12/2020, 16:13:55] ^[[39mLoaded plugin: homebridge-z2m@1.0.2
^[[37m[31/12/2020, 16:13:55] ^[[39mRegistering platform 'homebridge-z2m.zigbee2mqtt'
^[[37m[31/12/2020, 16:13:55] ^[[39m---
^[[37m[31/12/2020, 16:13:55] ^[[39mLoading 3 platforms...

^[[37m[31/12/2020, 16:13:55] ^[[39m^[[36m[Config] ^[[39mInitializing config platform...
^[[37m[31/12/2020, 16:13:55] ^[[39m^[[36m[Config] ^[[39mRunning in Service Mode
^[[37m[31/12/2020, 16:13:55] ^[[39m^[[36m[Intercom] ^[[39mInitializing IntercomAutomationHAT platform...
^[[37m[31/12/2020, 16:13:55] ^[[39m^[[36m[Intercom] ^[[39mIntercomPlatform constructor
^[[37m[31/12/2020, 16:13:55] ^[[39m^[[36m[Intercom] ^[[39mSetting up pyhton shell. scriptPath = /usr/local/lib/node_modules/homebri$^[[37m[31/12/2020, 16:13:56] ^[[39m^[[36m[Intercom] ^[[39mCreating IntercomPlatform accessories...
^[[37m[31/12/2020, 16:13:56] ^[[39m^[[36m[Intercom] ^[[39mIntercomLockAccessory constructor
^[[37m[31/12/2020, 16:13:56] ^[[39m^[[36m[Intercom] ^[[39mprepareLockAccessory
^[[37m[31/12/2020, 16:13:56] ^[[39m^[[36m[Intercom] ^[[39mprepareLockService
^[[37m[31/12/2020, 16:13:56] ^[[39m^[[36m[Intercom] ^[[39mIntercomBellAccessory constructor
^[[37m[31/12/2020, 16:13:56] ^[[39m^[[36m[Intercom] ^[[39mprepareBellAccessory
^[[37m[31/12/2020, 16:13:56] ^[[39m^[[36m[Intercom] ^[[39mprepareDoorbellService
^[[37m[31/12/2020, 16:13:56] ^[[39m^[[36m[Intercom] ^[[39mconnectListingToDoorbel
^[[37m[31/12/2020, 16:13:56] ^[[39mInitializing platform accessory 'IntercomLock'...
^[[37m[31/12/2020, 16:13:56] ^[[39mInitializing platform accessory 'IntercomBell'...
^[[37m[31/12/2020, 16:13:56] ^[[39m^[[36m[zigbee2mqtt] ^[[39mInitializing zigbee2mqtt platform...
 ^[[37m[31/12/2020, 16:13:56] ^[[39m^[[36m[zigbee2mqtt] ^[[39mConnecting to MQTT server at mqtt://localhost:1883
Setup Payload:
[…]

^[[37m[31/12/2020, 16:13:59] ^[[39mHomebridge is running on port 51521.
^[[37m[31/12/2020, 16:14:00] ^[[39m^[[36m[zigbee2mqtt] ^[[39mConnected to MQTT server
^[[0;37m[31/12/2020, 16:16:48] ^[[0m^[[0;36m[Homebridge UI]^[[0m Changes to config.json saved.
`
And for v5:
``

**Comment by itavero** [comment]
Which MQTT server are you using? I'm running mosquitto v1.6.12.

If you check the logs of zigbee2mqtt, do you see the message to `zigbee2mqtt/bridge/devices` being published?

**Comment by tkrtz** [comment]
I'm running mosquitto 1.5.7 (as this is the newest for the Pi zero) and there isn't even a topic with `zigbee2mqtt/bridge/devices`.

**Comment by itavero** [comment]
@tkrtz I might have some time next week to setup a test environment in Docker with an older version of MQTT.
If the topic is not there, I think it's more likely that the issue is caused by some incompatibility with the newer versions of zigbee2mqtt and the version of mosquitto you are running.

Raspberry Pi Zero has an ARMv6 instruction set, right?
There are newer versions of the [eclipse-mosquitto container](https://hub.docker.com/_/eclipse-mosquitto) available on Docker Hub for ARMv6.

**Comment by tkrtz** [comment]
Okay, so installed the newer version of mosquitto (as the have published an ARMv6 build for version 1.6.12) and downgraded nodejs to v12.19.1.
Turns out the newest version of z2m had troubles installing while running the newest version of node. So by purging z2m and reinstalling it I‘ve gotten the `/bridge/devices` topic 🎉
All my zigbee devices are displayed in the HomeKit app!

Sorry to cause you troubles. At least I (or maybe we) know now, that using the latest version (or the recommended one) is quite important :)

And a happy new year!

**Comment by itavero** [comment]
@tkrtz Happy new year! Glad to hear it's working now.

**Comment by itavero** [comment]
Spend quite some time on #2 (support for remote controls) the last couple of days.
I think I'll be able to release that in the `next` channel as well in the coming weeks.

I still plan to check it I can make the migration from previous versions to this new major version better.
Unfortunately that means I have to setup a second Homebridge environment with the old version of the plugin, as I've been using the new major version at home since I published it.
I haven't ran into any major issues (yet), except for the migration from the old version (which is why I would want to improve on it if possible).

**Comment by itavero** [comment]
Just published [v1.1.0-beta.0](https://github.com/itavero/homebridge-z2m/blob/master/CHANGELOG.md#110-beta0---2021-01-10) on the [`next`](https://www.npmjs.com/package/homebridge-z2m/v/next) channel.

This adds support for push buttons (#2) and should improve the "migration" experience when coming from a pre-v1.0.0 version, as it tries to "recycle" existing services.

**Comment by itavero** [comment]
Closing this issue, as I feel that most "major" refactoring work is done.

If you run into issues with the latest beta, feel free to add a comment here or open up a new issue.

**Comment by Diederik98** [comment]
Hi @itavero,

I'm looking forward to test your beta version!
However I get an error that the dev branch of zigbee2mqtt is not supported. 
I'm using the v1.17.0-dev version atm. 

Is this something I can fix myself? (The dev branch is more stable for me...)

Thanks!

**Comment by itavero** [comment]
@Diederik98 Can you send me the error message? Then I can do some testing.

I think I need to make the [`checkZigbee2MqttVersion`](https://github.com/itavero/homebridge-z2m/blob/c732e63ae80257584d7ec990f45473693f682e0b/src/platform.ts#L111-L119) less strict.

Apparently `semver.lt` considers not having a prerelease greater than having one (which is correct, but is not how zigbee2mqtt is versioned).

**Comment by Diederik98** [comment]
@itavero Here you go
![Schermafbeelding 2021-01-12 om 09 26 06](https://user-images.githubusercontent.com/30418238/104288568-44356e00-54b8-11eb-8627-9e3cd68f4845.png)

Thanks!

**Comment by itavero** [comment]
Easy solution is probably to just ignore the `-dev` suffix if it's present. I'll see if I can add that as a quick workaround tonight.

**Comment by itavero** [comment]
@Diederik98 As you might have noticed I did not get around to it yet, unfortunately. Luckily it's the weekend now, so it _should_ be easier to find a moment to fix this. I'll let you know once the fix is available.

**Comment by itavero** [comment]
@Diederik98 The fix is available in v1.1.0-beta.1, available on the `next` channel (so `npm i homebridge-z2m@next` should fix things for you) 😉

**Comment by Diederik98** [comment]
@itavero Thanks! The small ikea remotes (E1743) are nice to have for manual toggles. Nice work!  🎉 I'm glad homekit shows which button has been pressed for the automations 😅

**Comment by itavero** [comment]
Indeed. You can also see the mapping in the Homebridge logs, by the way.
